tox.ini

```diff
@@ -1,15 +1,25 @@
 [tox]
-envlist = py3, flake8, pycodestyle
+envlist = py3, flake8, flake8-{trusty,xenial,bionic}
+
+# N.B. If this needs to be expanded to include a third type of test, a
+# "new-test: new-test-commands" line should be added to the commands stanza
+# (and perhaps the deps stanza, if additional dependencies are required), and
+# the default "!flake8" line should be extended to "!flake8-!new-test"
 
 [testenv]
 deps =
     -rrequirements.txt
     -rtest-requirements.txt
-commands = nose2 --with-coverage --coverage uaclient/ uaclient
-
-[testenv:flake8]
-commands = flake8 uaclient setup.py
-
-[testenv:pycodestyle]
-commands = pycodestyle {posargs:uaclient setup.py}
+    trusty: -cdev/trusty-constraints.txt
+    xenial: -cdev/xenial-constraints.txt
+    bionic: -cdev/bionic-constraints.txt
+commands =
+    !flake8: nose2 --with-coverage --coverage uaclient/ uaclient
+    flake8: flake8 uaclient setup.py
```

Comment:
This didn't work for me with tox 2.5.0: `ERROR: InvocationError: could not find executable '!flake8:'` I'm not sure why the conditional formatting wasn't working.

In either case shouldn't the first line by "py3:" instead so it only runs on py3 and that way it also simplifies your N.B. comment for adding new tests?

Reply:
Yep, looks like the negative conditional factors were introduced in [this commit](https://github.com/tox-dev/tox/commit/2ce35c0020ff22401ec7b8207032f5f2257aa6f3) and so are only available from 3.0 onwards. The docs suggest that all the conditional factor syntax was introduced in 1.8 (which would be fine everywhere but trusty, and we don't need to run tox on trusty); I've filed https://github.com/tox-dev/tox/issues/1201 to address that.

I will switch to the syntax you suggested.

(I used `!flake8` so invocations with other Python versions (e.g. `py34`, `py37`, `py2`) would work. This would only really be necessary if we needed to support Python 2, which we don't. If we want other such invocations to work, we can explicitly add them in future. For reference, `tox -e py37` with the new `tox.ini` gives: `ERROR: Commands not specified. Please update relevant section of <snip>/ubuntu-advantage-client/tox.ini`.)